paperqa/clients/openalex.py

```diff
@@ -0,0 +1,238 @@
+        title_similarity_threshold: The threshold for title similarity.
+
+    Returns:
+        The details of the document if found, otherwise None.
+
+    Raises:
+        ValueError: If neither DOI nor title is provided.
+        DOINotFoundError: If the paper cannot be found.
+    """
+    mailto = await get_openalex_mailto()
+    params = {"mailto": mailto} if mailto else {}
+
+    if doi is None and title is None:
+        raise ValueError("Either a DOI or title must be provided.")
+
+    url = f"{OPENALEX_BASE_URL}/works"
+    if doi:
+        url += f"/https://doi.org/{quote(doi, safe='')}"
+    elif title:
+        params["filter"] = f"title.search:{title}"
+
+    if fields:
+        params["select"] = ",".join(fields)
+
+    async with session.get(
+        url,
+        params=params,
+        timeout=aiohttp.ClientTimeout(OPENALEX_API_REQUEST_TIMEOUT),
+    ) as response:
+        try:
+            response.raise_for_status()
+            response_data = await response.json()
+        except (aiohttp.ClientResponseError, json.JSONDecodeError) as exc:
+            raise DOINotFoundError("Could not find paper given DOI/title.") from exc
+
+        if response_data.get("status") == "failed":
+            raise DOINotFoundError(
+                "OpenAlex API returned a failed status for the query."
+            )
+
+        results_data = response_data
+        if params.get("filter") is not None:
+            results_data = results_data["results"]
+            if len(results_data) == 0:
+                raise DOINotFoundError(
+                    "OpenAlex API did not return any items for the query."
+                )
+            results_data = results_data[0]
+
+        if (
+            doi is None
+            and title
+            and strings_similarity(results_data.get("title", ""), title)
+            < title_similarity_threshold
+        ):
+            raise DOINotFoundError(
+                f"OpenAlex results did not match for title {title!r}."
+            )
+        if doi and results_data.get("doi") != doi:
+            raise DOINotFoundError("DOI not found in OpenAlex")
+
+        return await parse_openalex_to_doc_details(results_data)
+
+
+async def parse_openalex_to_doc_details(message: dict[str, Any]) -> DocDetails:
+    """Parse OpenAlex API response to DocDetails.
+
+    Args:
+        message: The OpenAlex API response message.
+
+    Returns:
+        Parsed document details.
+    """
+
+    # author_name will be FamilyName, GivenName Middle initial. (if available)
+    # there is no standalone "FamilyName" or "GivenName" fields
+    # this manually constructs the name into the format the other clients use
+    def reformat_name(name: str) -> str:
+        # https://regex101.com/r/74vR57/1
+        pattern = r"^([^,]+),\s*(.+?)(?:\s+(\w+\.?))?$"
+        match = re.match(pattern, name)
+        if match:
+            family_name, given_name, middle = match.groups()
```

Comment:
Not going to work on second first names, chinese character parenthetical names, and suffixes. Some examples:

```text
Given A. Family
Foo-bar B. Family
Given Second A. Family, Sr.
Foo Bar(我的名字)
```

Probably need to do something bone headed like:
```python
*firsts, last = names.split()
first_name = " ".join(firsts)
```

Reply:
this should now be addressed. (added UTs for this specific function)